userland/libtock/Makefile

```diff
@@ -5,45 +5,60 @@ ifeq ($(LIBTOCK_DIR),/libtock)
 endif
 
 $(call check_defined, TOCK_USERLAND_BASE_DIR)
-$(call check_defined, TOCK_ARCH)
+$(call check_defined, TOCK_ARCHS)
 $(call check_defined, AR)
 $(call check_defined, CC)
 $(call check_defined, RANLIB)
 
-LIBTOCK_BUILDDIR ?= $(LIBTOCK_DIR)/build/$(TOCK_ARCH)
+LIBTOCK_BUILDDIR ?= $(LIBTOCK_DIR)/build
 
 CFLAGS += -I$(LIBTOCK_DIR)
 
 LIBTOCK_C_SRCS  := $(wildcard $(LIBTOCK_DIR)/*.c)
 LIBTOCK_AS_SRCS := $(wildcard $(LIBTOCK_DIR)/*.s)
-LIBTOCK_OBJS    := $(patsubst $(LIBTOCK_DIR)/%.s,$(LIBTOCK_BUILDDIR)/%.o,$(LIBTOCK_AS_SRCS))
-LIBTOCK_OBJS    += $(patsubst $(LIBTOCK_DIR)/%.c,$(LIBTOCK_BUILDDIR)/%.o,$(LIBTOCK_C_SRCS))
 
 .PHONY: all
-all: $(LIBTOCK_BUILDDIR)/libtock.a
+all: $(foreach arch, $(TOCK_ARCHS),$(LIBTOCK_BUILDDIR)/$(arch)/libtock.a)
 
-$(LIBTOCK_BUILDDIR):
-	$(Q)mkdir -p $@
+# Rules to generate LibTock for a given Architecture
+# These will be used to create the different architecture versions of LibTock
+# Argument $(1) is the Architecture (e.g. cortex-m0) to build for
+#
+# To see the generated rules, run:
+# $(info $(foreach arch,$(TOCK_ARCHS),$(call LIBTOCK_RULES,$(arch))))
+define LIBTOCK_RULES
 
-$(LIBTOCK_BUILDDIR)/%.o: $(LIBTOCK_DIR)/%.c | $(LIBTOCK_BUILDDIR)
-	$(TRACE_DEP)
-	$(Q)$(CC) $(CFLAGS) $(CPPFLAGS) -MF"$(@:.o=.d)" -MG -MM -MP -MT"$(@:.o=.d)@" -MT"$@" "$<"
-	$(TRACE_CC)
-	$(Q)$(CC) $(CFLAGS) $(CPPFLAGS) -c -o $@ $<
+$(LIBTOCK_BUILDDIR)/$(1):
+	$(Q)mkdir -p $$@
 
-$(LIBTOCK_BUILDDIR)/%.o: %.S | $(LIBTOCK_BUILDDIR)
-	$(TRACE_AS)
-	$(Q)$(AS) $(ASFLAGS) $(CPPFLAGS) -c -o $@ $<
+$(LIBTOCK_BUILDDIR)/$(1)/%.o: $(LIBTOCK_DIR)/%.c | $(LIBTOCK_BUILDDIR)/$(1)
+	$$(TRACE_DEP)
+	$(Q)$(CC) $(CFLAGS) -mcpu=$(1) $(CPPFLAGS) -MF"$$(@:.o=.d)" -MG -MM -MP -MT"$$(@:.o=.d)@" -MT"$$@" "$$<"
+	$$(TRACE_CC)
+	$(Q)$(CC) $(CFLAGS) -mcpu=$(1) $(CPPFLAGS) -c -o $$@ $$<
 
-$(LIBTOCK_BUILDDIR)/libtock.a: $(LIBTOCK_OBJS) | $(LIBTOCK_BUILDDIR)
-	$(TRACE_AR)
-	$(Q)$(AR) rc $@ $^
-	$(Q)$(RANLIB) $@
+$(LIBTOCK_BUILDDIR)/$(1)/%.o: %.S | $(LIBTOCK_BUILDDIR)/$(1)
+	$$(TRACE_AS)
+	$(Q)$(AS) $(ASFLAGS) -mcpu=$(1) $(CPPFLAGS) -c -o $$@ $$<
+
+#XXX: It would be way better if we could somehow define LIBTOCK_OBJS again instead of just copying those lines everywhere
+#LIBTOCK_OBJS    := $(patsubst $(LIBTOCK_DIR)/%.s,$(LIBTOCK_BUILDDIR)/%.o,$(LIBTOCK_AS_SRCS))
+#LIBTOCK_OBJS    += $(patsubst $(LIBTOCK_DIR)/%.c,$(LIBTOCK_BUILDDIR)/%.o,$(LIBTOCK_C_SRCS))
+
+$(LIBTOCK_BUILDDIR)/$(1)/libtock.a: $(patsubst $(LIBTOCK_DIR)/%.s,$(LIBTOCK_BUILDDIR)/$(1)/%.o,$(LIBTOCK_AS_SRCS)) $(patsubst $(LIBTOCK_DIR)/%.c,$(LIBTOCK_BUILDDIR)/$(1)/%.o,$(LIBTOCK_C_SRCS)) | $(LIBTOCK_BUILDDIR)/$(1)
```

Comment:
same `patsubst` comment here, but from your comment above it seems like you tried?